src/device-state.coffee

```diff
@@ -0,0 +1,512 @@
+	if state.name? and !validation.isValidShortText(state.name)
+		throw new Error('Invalid device name')
+	if state.apps? and !validation.isValidAppsObject(state.apps)
+		throw new Error('Invalid apps')
+	if state.config? and !validation.isValidEnv(state.config)
+		throw new Error('Invalid device configuration')
+
+validateDependentState = (state) ->
+	if state.apps? and !validation.isValidDependentAppsObject(state.apps)
+		throw new Error('Invalid dependent apps')
+	if state.devices? and !validation.isValidDependentDevicesObject(state.devices)
+		throw new Error('Invalid dependent devices')
+
+validateState = Promise.method (state) ->
+	throw new Error('State must be an object') if !_.isObject(state)
+	validateLocalState(state.local) if state.local?
+	validateDependentState(state.dependent) if state.dependent?
+
+class DeviceStateRouter
+	constructor: (@deviceState) ->
+		{ @applications } = @deviceState
+		@router = express.Router()
+		@router.use(bodyParser.urlencoded(extended: true))
+		@router.use(bodyParser.json())
+
+		@router.post '/v1/reboot', (req, res) =>
+			force = validation.checkTruthy(req.body.force)
+			@deviceState.executeStepAction({ action: 'reboot' }, { force })
+			.then (response) ->
+				res.status(202).json(response)
+			.catch (err) ->
+				if err instanceof updateLock.UpdatesLockedError
+					status = 423
+				else
+					status = 500
+				res.status(status).json({ Data: '', Error: err?.message or err or 'Unknown error' })
+
+		@router.post '/v1/shutdown', (req, res) =>
+			force = validation.checkTruthy(req.body.force)
+			@deviceState.executeStepAction({ action: 'shutdown' }, { force })
+			.then (response) ->
+				res.status(202).json(response)
+			.catch (err) ->
+				if err instanceof updateLock.UpdatesLockedError
+					status = 423
+				else
+					status = 500
+				res.status(status).json({ Data: '', Error: err?.message or err or 'Unknown error' })
+
+		@router.get '/v1/device', (req, res) =>
+			@deviceState.getStatus()
+			.then (state) ->
+				stateToSend = _.pick(state.local, [
+					'api_port'
+					'commit'
+					'ip_address'
+					'status'
+					'download_progress'
+					'os_version'
+					'supervisor_version'
+					'update_pending'
+					'update_failed'
+					'update_downloaded'
+				])
+				res.json(stateToSend)
+			.catch (err) ->
+				res.status(500).json({ Data: '', Error: err?.message or err or 'Unknown error' })
+
+		@router.use(@applications.router)
+
+module.exports = class DeviceState extends EventEmitter
+	constructor: ({ @db, @config, @eventTracker }) ->
+		@logger = new Logger({ @eventTracker })
+		@deviceConfig = new DeviceConfig({ @db, @config, @logger })
+		@applications = new ApplicationManager({ @config, @logger, @db, @eventTracker })
+		@on 'error', (err) ->
+			console.error('Error in deviceState: ', err, err.stack)
+		@_currentVolatile = {}
+		_lock = new Lock()
+		@_writeLock = Promise.promisify(_lock.async.writeLock)
+		@_readLock = Promise.promisify(_lock.async.writeLock)
```

Comment:
Did you mean `_lock.async.readLock` here?

Reply:
Ah, yeah, fixed